### PR TITLE
Wallet-based, encrypted key management

### PR DESCRIPTION
**Pull request by paulhauner**
## Issue Addressed

NA

## Proposed Changes

Removes the old unencrypted `SszEncodableKeypair` method of storing validator keys and replaces it with an approach based upon:

- Encrypted keystores via EIP-2335.
- HD key derivation via EIP-2333.
- Encrypted wallets via EIP-2386 (draft).

## Additional Info

- ~~Blocked on #879~~

## TODO:

- [x] Pre-generate wallets for sim (wontfix: can do this later, sim runs in reasonable time without this)
- [x] Add new logic for deposits
  - [x] Remove old `validator_directory` mod and `account validator ...` commands
- [x] Add tests for `account_manager/src/wallet/...`
- [x] Add docs+tests for `validator_dir`.
- [x] Add some docs to the top of `lcli/src/upgrade_legacy_keypairs.rs`
- ~~[ ] Fix no-eth1 sim~~
- [x] Update docs
- [x] Require .pass suffix on CLI.
- [x] Resolve TODO about Mnemonic generation.

Issues to create:

- [x] Add tests for deposit CLI.
- [x] Remove legacy keypair stuff
- [x] Upgrade local testnet scrips
- [x] Make PR to lighthouse-docker
- [x] Eth1 Sim

### CODE REVIEW

File: validator_client/src/lib.rs [hunk 1]
```diff
@@ -159,28 +157,14 @@ impl<T: EthSpec> ProductionValidatorClient<T> {
...
-                config.data_dir.clone(),
-                genesis_validators_root,
-                context.eth2_config.spec.clone(),
-                fork_service.clone(),
-                log.clone(),
-            )?,
-            // Generate ephemeral insecure keypairs for testing purposes.
-            //
-            // Do not use in production.
-            KeySource::InsecureKeypairs(indices) => ValidatorStore::insecure_ephemeral_validators(
```
Comment (michaelsproul) [inline validator_client/src/lib.rs:176]: Yay for getting rid of the `InsecureKeypairs` stuff

File: tests/simulator/src/no_eth1_sim.rs [hunk 2]
```diff
@@ -72,37 +93,37 @@ pub fn run_no_eth1_sim(matches: &ArgMatches) -> Result<(), String> {
...
         /*
-         * One by one, add validator clients to the network. Each validator client is attached to
-         * a single corresponding beacon node.
+         * Create a future that will add validator clients to the network. Each validator client is
+         * attached to a single corresponding beacon node.
          */
+        let add_validators_fut = async {
+            for (i, files) in validator_files.into_iter().enumerate() {
+                network
+                    .add_validator_client(ValidatorConfig::default(), i, files)
```
Comment (michaelsproul) [inline tests/simulator/src/no_eth1_sim.rs:104]: This config will need to be:

```rust
ValidatorConfig {
                        auto_register: true,
                        ..ValidatorConfig::default()
}
```

File: tests/simulator/src/eth1_sim.rs [hunk 3]
```diff
@@ -103,54 +124,57 @@ pub fn run_eth1_sim(matches: &ArgMatches) -> Result<(), String> {
...
         /*
-         * One by one, add validator clients to the network. Each validator client is attached to
-         * a single corresponding beacon node.
+         * Create a future that will add validator clients to the network. Each validator client is
+         * attached to a single corresponding beacon node.
          */
+        let add_validators_fut = async {
+            for (i, files) in validator_files.into_iter().enumerate() {
+                network
+                    .add_validator_client(ValidatorConfig::default(), i, files)
```
Comment (michaelsproul) [inline tests/simulator/src/eth1_sim.rs:135]: Same as before, this config needs to include `auto_register: true`

File: eth2/utils/eth2_wallet_manager/src/locked_wallet.rs [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+        if !wallet_dir.exists() {
+            return Err(Error::MissingWalletDir(wallet_dir));
+        }
+
+        let lockfile = wallet_dir.join(LOCK_FILE);
+        if lockfile.exists() {
+            return Err(Error::WalletIsLocked(wallet_dir));
+        } else {
+            File::create(lockfile).map_err(Error::UnableToCreateLockfile)?;
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply eth2/utils/eth2_wallet_manager/src/locked_wallet.rs:56]: Great pickup on these! I _knew_ I was doing something wrong here!

**Comment by paulhauner** [comment]
All comments addressed, thank you!

**Pull request merged**